clients/apply/source.go

```diff
@@ -0,0 +1,215 @@
+package apply
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/influxdata/influx-cli/v2/api"
+	"github.com/influxdata/influx-cli/v2/pkg/github"
+	"github.com/influxdata/influx-cli/v2/pkg/jsonnet"
+	"gopkg.in/yaml.v3"
+)
+
+type TemplateEncoding int
+
+const (
+	TemplateEncodingUnknown TemplateEncoding = iota
+	TemplateEncodingJson
+	TemplateEncodingJsonnet
+	TemplateEncodingYaml
+)
+
+func (e *TemplateEncoding) Set(v string) error {
+	switch v {
+	case "jsonnet":
+		*e = TemplateEncodingJsonnet
+	case "json":
+		*e = TemplateEncodingJson
+	case "yml", "yaml":
+		*e = TemplateEncodingYaml
+	default:
+		return fmt.Errorf("unknown inEncoding %q", v)
+	}
+	return nil
+}
+
+func (e TemplateEncoding) String() string {
+	switch e {
+	case TemplateEncodingJsonnet:
+		return "jsonnet"
+	case TemplateEncodingJson:
+		return "json"
+	case TemplateEncodingYaml:
+		return "yaml"
+	case TemplateEncodingUnknown:
+		fallthrough
+	default:
+		return "unknown"
+	}
+}
+
+type TemplateSource struct {
+	Name     string
+	Encoding TemplateEncoding
+	Open     func(context.Context) (io.ReadCloser, error)
+}
+
+func SourcesFromPath(path string, recur bool, encoding TemplateEncoding) ([]TemplateSource, error) {
+	paths, err := findPaths(path, recur)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find inputs at path %q: %w", path, err)
+	}
+
+	sources := make([]TemplateSource, len(paths))
+	for i := range paths {
+		path := paths[i] // Local var for the `Open` closure to capture.
+		encoding := encoding
+		if encoding == TemplateEncodingUnknown {
+			ext := filepath.Ext(path)
+			switch {
+			case strings.HasPrefix(ext, ".jsonnet"):
+				encoding = TemplateEncodingJsonnet
+			case strings.HasPrefix(ext, ".json"):
+				encoding = TemplateEncodingJson
+			case strings.HasPrefix(ext, ".yml") || strings.HasPrefix(ext, ".yaml"):
+				encoding = TemplateEncodingYaml
+			default:
+			}
+		}
+
+		sources[i] = TemplateSource{
+			Name:     path,
+			Encoding: encoding,
+			Open: func(context.Context) (io.ReadCloser, error) {
+				return os.Open(path)
+			},
+		}
+	}
+	return sources, nil
+}
+
+func findPaths(path string, recur bool) ([]string, error) {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return nil, err
+	}
+	if !fi.IsDir() {
+		return []string{path}, nil
+	}
+
+	dirFiles, err := os.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var paths []string
+	for _, df := range dirFiles {
+		fullPath := filepath.Join(path, df.Name())
+		if df.IsDir() && recur {
+			subPaths, err := findPaths(fullPath, recur)
+			if err != nil {
+				return nil, err
+			}
+			paths = append(paths, subPaths...)
+		} else if !df.IsDir() {
+			paths = append(paths, fullPath)
+		}
+	}
+	return paths, nil
+}
+
+func SourceFromURL(u *url.URL, encoding TemplateEncoding) TemplateSource {
+	if encoding == TemplateEncodingUnknown {
+		ext := path.Ext(u.Path)
+		switch {
```

Comment:
I think I see why the switch statement above using the `HasPrefix` - it uses it here too for urls, which I imagine might have query parameters or other stuff (?) after the file name in the url. Makes sense to have them be the same if that's the reasoning.

Reply:
I wouldn't expect any query parameters to show up in the post-parsing `u.Path`. I'll refactor to drop the prefix-checks